File: interpreter-for-cpp/src/Runtime/Space.cpp
#include "Space.h"
#include "System.h"
#include "Value/Value.h"
#include "Variable.h"

using namespace peak::interpreter;

Space::Space(SpaceType spaceType)
	: _spaceType(spaceType) {
}
Space::Space(SpaceType spaceType, std::shared_ptr<Space> parent)
	: _spaceType(spaceType), _parent(parent) {
}

std::shared_ptr<Space> Space::CopySpace() const {
	std::shared_ptr<Space> parent{nullptr};
	if (_parent) {
		parent = _parent->CopySpace();
	}
	auto space = std::shared_ptr<Space>(new Space(_spaceType, parent));
	space->_spaceOfUsing = _spaceOfUsing;

	for (auto& pair : _variables) {
		auto tempVariable = pair.second;
		auto tempVlaue = tempVariable->GetValue();
		auto variable = std::shared_ptr<Variable>(new Variable(pair.first, tempVariable->GetAttribute()));
		if (tempVlaue) {
			variable->SetValue(tempVlaue->Clone());
		}
		space->_variables.emplace(pair.first, variable);
	}
	return space;
}

void Space::Clear() {
	_variables.clear();
}

bool Space::AddVariable(std::shared_ptr<Variable> value) {
	if (!value) {
		return false;
	}
	if (_variables.find(value->GetName()) != _variables.end()) {
		ErrorLogger::LogRuntimeError(value->GetName());
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Space, "The variable \"" + value->GetName() + "\" is exist!");
		return false;
	}
	_variables.emplace(value->GetName(), value);
	return true;
}

std::shared_ptr<Variable> Space::FindVariableFromTop(const std::string& name) const {
	auto ite = _variables.find(name);
	if (ite != _variables.end()) {
		return ite->second;
	}
	return nullptr;
}

std::shared_ptr<Variable> Space::FindVariable(const std::string& name) const {
	auto ite = _variables.find(name);
	if (ite != _variables.end()) {
		return ite->second;
	}
	for (auto space : _spaceOfUsing) {
		auto find = space->FindVariable(name);
		if (find) {
			return find;
		}
	}
	if (_parent) {
		return _parent->FindVariable(name);
	}
	return System::FindVariable(name);
}

SpaceType Space::GetSpaceType() const {
	return _spaceType;
}

void Space::AddSpaceOfUsing(std::shared_ptr<Space> space) {
	_spaceOfUsing.emplace_back(space);
}

File: interpreter-for-cs/src/Runtime/Space.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System.Collections.Generic;

namespace peak.interpreter {

	public enum SpaceType {
		None,
		Condition,
		Loop,
		Function,
		Object,
	}

	public class Space {
		public static implicit operator bool(Space value) {
			return value != null;
		}
		private SpaceType _spaceType = SpaceType.None;
		private Space _parent = null;
		private List<Space> _spaceOfUsing = new List<Space>();
		private Dictionary<string, Variable> _variables = new Dictionary<string, Variable>();

		public SpaceType spaceType { get => _spaceType; }

		public Space(SpaceType spaceType) {
			_spaceType = spaceType;
		}
		public Space(SpaceType spaceType, Space parent) {
			_spaceType = spaceType;
			_parent = parent;
		}

		public Space CopySpace() {
			Space parent = null;
			if (_parent) {
				parent = _parent.CopySpace();
			}
			var space = new Space(_spaceType, parent);
			space._spaceOfUsing = _spaceOfUsing;

			foreach (var pair in _variables) {
				var tempVariable = pair.Value;
				var tempVlaue = tempVariable.value;
				var variable = new Variable(pair.Key, tempVariable.attribute, tempVlaue ? tempVlaue.Clone() : null);
				space._variables.Add(pair.Key, variable);
			}
			return space;
		}
		public void Clear() {
			_variables.Clear();
		}

		public bool AddVariable(Variable value) {
			if (!value) {
				return false;
			}
			if (_variables.ContainsKey(value.name)) {
				ErrorLogger.LogRuntimeError(value.name);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Space, "The variable \"" + value.name + "\" is exist!");
				return false;
			}
			_variables.Add(value.name, value);
			return true;
		}
		public Variable FindVariable(string name) {
			Variable ret = null;
			if (_variables.TryGetValue(name, out ret)) {
				return ret;
			}
			foreach (var space in _spaceOfUsing) {
				ret = space.FindVariable(name);
				if (ret) {
					return ret;
				}
			}
			if (_parent) {
				return _parent.FindVariable(name);
			}
			return BuiltIn.instance.FindVariable(name);
		}
		public Variable FindVariableFromTop(string name) {
			Variable ret = null;
			if (_variables.TryGetValue(name, out ret)) {
				return ret;
			}
			return null;
		}

		public void AddSpaceOfUsing(Space space) {
			_spaceOfUsing.Add(space);
		}
	}
} // namespace peak.interpreter